### DISCUSSION TITLE
ToolDev with Concourse and go

### DISCUSSION THREAD

**Discussion by dio-trinny** [concourse/concourse#6312] [Help & Support] [answered]
Hello all!

I am trying to write a new cli tool to interact with concourse without having to tell my users to install fly, just this cli tool. I've run into a bit of an issue with go mod however, and I am trying to work out what it is that I am doing wrong. In particular, building this go file:

```
package main

import (
	"github.com/concourse/fly/commands"
)

func main() {
	fl := commands.Fly
	fl.Target = "whatever"

	lgCmd := commands.LoginCommand{
		ATCURL: "whatever",
	}
	err := lgCmd.Execute([]string{})
	if err != nil {
		panic(err)
	}
}
```

Results in these errors:
```
$ go mod tidy
go: finding module for package github.com/concourse/skymarshal/skycmd
go: finding module for package github.com/concourse/skymarshal
go: finding module for package k8s.io/api/auditregistration/v1alpha1
go: finding module for package github.com/concourse/web
github.com/mypackage imports
	github.com/concourse/fly/commands imports
	github.com/concourse/skymarshal/skycmd: module github.com/concourse/skymarshal@latest found (v0.0.0-20190409200102-eed82ea4219a), but does not contain package github.com/concourse/skymarshal/skycmd
github.com/mypackage imports
	github.com/concourse/fly/commands imports
	github.com/concourse/atc tested by
	github.com/concourse/atc.test imports
	github.com/concourse/atc/atccmd imports
	github.com/concourse/skymarshal: module github.com/concourse/skymarshal@latest found (v0.0.0-20190409200102-eed82ea4219a), but does not contain package github.com/concourse/skymarshal
github.com/mypackage imports
	github.com/concourse/fly/commands imports
	github.com/concourse/atc tested by
	github.com/concourse/atc.test imports
	github.com/concourse/atc/atccmd imports
	github.com/concourse/web: module github.com/concourse/web@latest found (v0.0.0-20190409200112-645ef4eb43bc), but does not contain package github.com/concourse/web
github.com/mypackage imports
	github.com/concourse/fly/commands imports
	github.com/concourse/atc tested by
	github.com/concourse/atc.test imports
	github.com/concourse/atc/atccmd imports
	github.com/concourse/atc/creds/kubernetes imports
	k8s.io/client-go/kubernetes imports
	k8s.io/client-go/kubernetes/typed/auditregistration/v1alpha1 imports
	k8s.io/api/auditregistration/v1alpha1: module k8s.io/api@latest found (v0.19.4), but does not contain package k8s.io/api/auditregistration/v1alpha1
```

I have checked that this package is actually shipped with concourse releases, so I am kind of lost.. Any help would be greatly appreciated!

**Comment by chenbh** [comment, marked as answer]
The [`concourse/fly`](https://github.com/concourse/fly) repo is archived and has no code in it, you should be using the `concourse/concourse` mono-repo instead. You might also want to take a look at using the [go-concourse client library](https://godoc.org/github.com/concourse/concourse/go-concourse/concourse#Client) instead of fly.

Note: due to how go mod versioning works, you'll have to get the repo by sha: `go get github.com/concourse/concourse@248606d` for v6.7.1. Eventually we'll separate out go-concourse into a proper go mod versioned repo, but for now you'll have to `go get` it by sha.

**Reply by dio-trinny** [reply]
thank you for your reply @chenbh ! I am trying to perform the Login and SetPipeline operations that fly allows you to, but go-concourse doesn't seem to export it.

What would be the best way of going about it in your opinion?!

**Reply by chenbh** [reply]
Logging in is done at the http.Client level, you can find examples of it at https://github.com/concourse/concourse/blob/master/fly/rc/target.go#L155 or https://github.com/alphagov/terraform-provider-concourse/blob/master/pkg/client/client.go#L35

Pipeline operations are at the [team](https://godoc.org/github.com/concourse/concourse/go-concourse/concourse#Team) level
